kernel/src/scan/partition_skipping.rs

```
@@ -95,33 +94,39 @@ impl RowVisitor for PartitionVisitor {
                     .iter()
+                    .filter(|(k, _v)| self.schema.field(k).is_some())
                     .map(|(k, v)| {
-                        let data_type = self.schema.field(k).map(|f| f.data_type());
-                        let primitve_type = if let Some(DataType::Primitive(primitive)) = data_type
-                        {
-                            primitive.to_owned()
-                        } else {
-                            return Err(crate::Error::Generic(
-                                "partition filtering only supported for primitive types"
-                                    .to_string(),
+                        // The schema we are evaluating only contains the fields of interest,
+                        // not all partition fields. The unwrap is safe because we have already
+                        // checked in the filter above.
+                        let data_type = self.schema.field(k).unwrap().data_type();
+
+                        let DataType::Primitive(primitive_type) = data_type else {
+                            return Err(crate::Error::unsupported(
+                                format!("Partition filtering only supported for primitive types. Found type: {}", data_type)
                             ));
                         };
 
-                        let scalar = primitve_type.parse_scalar(v)?;
+                        let scalar = primitive_type.parse_scalar(v)?;
                         Ok((ColumnName::new([k]), scalar))
                     })
-                    .collect::<DeltaResult<HashMap<ColumnName, Scalar>>>()
-                    .ok()?;
+                    .collect::<DeltaResult<HashMap<ColumnName, Scalar>>>()?;
+
                 let filter = DefaultPredicateEvaluator::from(resolver);
-                Some(filter.eval_expr(&self.predicate, false).unwrap_or(true))
+                Ok(filter.eval_expr(&self.predicate, false).unwrap_or(false))
```

This doesn't look right? If the expression evaluated to `None` that means we don't know whether it can be skipped and we must _not_ skip it (see my other comment about the importance of passing `Scalar::Null` values for partition values that are known to be NULL, so that IS [NOT] NULL predicates work correctly).

Meanwhile, if we just need to handle the extra Option/Result nesting, we can do:
```rust
let val = getter.get_map(i, "output")?.and_then(|m| {
      ...
    Ok(filter.eval_expr(&self.predicate, false)).transpose()
});
```

---

I see your point now. 

Something I was struggling with is I was testing with the multi partitioned acceptance tests examples, some of which do have null values for some partitions. In particular looking at `acceptance/tests/dat/out/reader_tests/generated/multi_partitioned` I was trying to use a predicate that column "letter" equals a literal "b". This does let the nulls through because in the equality we're using `PartialCmp` even for equality. I get that when doing a > or < comparison it's ill defined for null values, but I would expect that checking equality should be valid. I tested adding in a `partial_eq_scalars` along side `partial_cmp_scalars` and it resolves the problem I was having.

I might coming at the problem from a different point of view than the delta-kernel design. It the intent is for nulls to make it through this kind of predicate, then I can adjust my datafusion code to add in additional predicates of "is not null" for each of these cases.

---

Comparisons with NULL get tricky... in SQL, any comparison against NULL produces NULL -- including e.g. `NULL == NULL`. There's a concept of "null-safe equal" in spark (operator `<=>`) which always returns either TRUE or FALSE, and there are other ways in other systems to achieve a similar result. But yes -- if you want a null-safe equal (or comparison) in a system that doesn't explicitly claim to support it, you need to turn e.g. `x > 10` into `x IS NOT NULL AND x > 10`. 

Additionally for our predicate filters, a missing value is as NULL some of the time (e.g. comparisons against it return NULL), but it's _NOT_ the same when it comes to IS [NOT] NULL predicates. Because if a value is outright missing, we don't know whether that's because it was null, or because it was unavailable. So a missing input has to cause IS [NOT] NULL to return NULL, when it comes to data skipping. For other comparisons, the correct thing already happens naturally because the result is the same for missing vs. NULL. 

To give an example, suppose the predicate was:
```sql
WHERE partition_col > 10 AND value_col = 42
```
Then for partition pruning, we won't have stats for `value_col` and we'll effectively end up with:
```sql
WHERE partition_col > 10 AND NULL
```
That's totally fine -- `FALSE AND NULL` is `FALSE`, so we can still skip. `TRUE AND NULL` is `NULL`, but that's fine because `TRUE` and `NULL` both force us to keep the file. 

Complicated stuff! For more details, see extensive code comments including:
https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/predicates/mod.rs#L25-L44
https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/predicates/parquet_stats_skipping.rs#L109-L153

---

I should add, data skipping and SQL WHERE have conflicting requirements:
* Data skipping must _keep_ the file _unless_ the predicate returns an actual FALSE
* WHERE _discards_ the row _unless_ the predicate returns an actual TRUE

For partition pruning, we should in theory use SQL WHERE semantics -- but that's only safe if we're certain that _every_ column the predicate references is a partition column, which is not guaranteed. Hence my compromise suggestion, that we should ensure the partition values map always has a value (`Scalar::Null` by default) for _every_ partition column.

Additionally, we will probably want to hoist the `eval_sql_where` function from `ParquetStatsSkippingFilter` (which can probably go away) up to the `PredicateEvaluator` trait so you can use it. That way, the required `IS NOT NULL` checks would be injected automatically.

---

Thank you for taking the time to write this all up. From reading the delta spec, specifically this line in [requirements for writers](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#requirements-for-writers) :

> Values for all partition columns present in the schema MUST be present for all files in the table.

This would lead me to think that the right thing to do is that if any of the `partitionValues` are entirely missing that we should return an error because it doesn't match the spec.

Also from my reading of the spec it looks like null values should be an empty string per [partition value serialization](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#partition-value-serialization). The acceptance tests currently have things like `"partitionValues":{"letter":"a","date":null,"data":"x"}` which is also how my other examples show up. But to meet the spec it looks like we need to also support the empty string case. I don't immediately see any examples in the repo that have this for a unit test, so I'll be sure to add both variants (`null` and `""`).

---

Hoisting `eval_sql_where` up so `DefaultPredicateEvaluator` can use it: https://github.com/delta-io/delta-kernel-rs/pull/627

---

> Thank you for taking the time to write this all up. From reading the delta spec, specifically this line in [requirements for writers](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#requirements-for-writers) :
> 
> > Values for all partition columns present in the schema MUST be present for all files in the table.
> 
> This would lead me to think that the right thing to do is that if any of the `partitionValues` are entirely missing that we should return an error because it doesn't match the spec.

That sounds right to me as well. However:
1. We may want to ask delta-rs folks if they've seen other behaviors in the wild that we need to consider tolerating.
2. We still have to worry about the non-partition columns the predicate might mention -- they won't have entries in `add.partitionValues` and we _want_ them to resolve as `None` (_not_ `Scalar::Null`).

Maybe the simplest thing to do -- in this partition skipping code at least -- is to treat missing partition values as if they were value columns. They would simply not participate in skipping, and the query that comes later can error out as appropriate.